src/humanloop/decorators/tool.py

```diff
@@ -0,0 +1,440 @@
+        func.json_schema = tool_kernel["function"]  # type: ignore
+
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            with opentelemetry_tracer.start_as_current_span(str(uuid.uuid4())) as span:
+                span_id = span.get_span_context().span_id
+                if span.parent:
+                    span_parent_id = span.parent.span_id
+                else:
+                    span_parent_id = None
+                parent_trace_metadata = TRACE_FLOW_CONTEXT.get(span_parent_id)
+                if parent_trace_metadata:
+                    TRACE_FLOW_CONTEXT[span_id] = FlowContext(
+                        span_id=span_id,
+                        trace_parent_id=span_parent_id,
+                        is_flow_log=False,
+                    )
+
+                # Write the Tool Kernel to the Span on HL_FILE_OT_KEY
+                span.set_attribute(HL_PATH_KEY, path if path else func.__name__)
+                span.set_attribute(HL_FILE_TYPE_KEY, "tool")
+                if tool_kernel:
+                    write_to_opentelemetry_span(
+                        span=span,
+                        key=f"{HL_FILE_KEY}.tool",
+                        value=tool_kernel,
+                    )
+
+                # Call the decorated function
+                try:
+                    output = func(*args, **kwargs)
+                    error = None
+                except Exception as e:
+                    logger.error(f"Error calling {func.__name__}: {e}")
+                    output = None
+                    error = str(e)
+
+                # Populate known Tool Log attributes
+                tool_log = {
+                    "inputs": args_to_inputs(func, args, kwargs),
+                    "output": output,
+                    "error": error,
+                }
+
+                # Write the Tool Log to the Span on HL_LOG_OT_KEY
+                write_to_opentelemetry_span(
+                    span=span,
+                    key=HL_LOG_KEY,
+                    value=tool_log,
+                )
+
+                # Return the output of the decorated function
+                return output
+
+        func.file = File(  # type: ignore
+            path=path if path else func.__name__,
+            type="tool",
+            version=tool_kernel,
+            is_decorated=True,
+            callable=wrapper,
+        )
+
+        return wrapper
+
+    return decorator
+
+
+def _build_tool_kernel(
+    func: Callable,
+    attributes: Optional[dict[str, Optional[Any]]],
+    setup_values: Optional[dict[str, Optional[Any]]],
+    strict: bool,
+) -> ToolKernelRequestParams:
+    """Build ToolKernelRequest object from decorated function."""
+    kernel = ToolKernelRequestParams(
+        source_code=textwrap.dedent(
+            # Remove the tool decorator from source code
+            inspect.getsource(func).split("\n", maxsplit=1)[1]
```

Comment:
(i'd suggest not trying to do this cleanup if we can't make it robust; i doubt we can make it robust)

also decorators can be used directly `func = decorate(func)` (instead of
```
@decorate
def func():
    pass
```
would that just make this throw an error if there's no newline in the `inspect.getsource()` result?

Reply:
The source code kernel parameter cannot be specified via the decorator. I have revisited that part of the decorator and made the extraction robust: using the decorator as intended or as a higher-order function will extract the same source code reliably. I have added a test for this, good catch.